internal/pkg/manifest/transform.go

```diff
@@ -19,89 +42,281 @@ var basicKinds = []reflect.Kind{
+	if typ != reflect.TypeOf(PlatformArgsOrString{}) {
+		return nil
+	}
+
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(PlatformArgsOrString), src.Interface().(PlatformArgsOrString)
+
+		if srcStruct.PlatformString != nil {
+			dstStruct.PlatformArgs = PlatformArgs{}
+		}
+
+		if !srcStruct.PlatformArgs.isEmpty() {
+			dstStruct.PlatformString = nil
+		}
+
+		if dst.CanSet() { // For extra safety to prevent panicking.
+			dst.Set(reflect.ValueOf(dstStruct))
+		}
+		return nil
+	}
 }
 
-type imageTransformer struct{}
+type healthCheckArgsOrStringTransformer struct{}
 
-// Transformer returns custom merge logic for volume's fields.
-func (t imageTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
-	if typ == reflect.TypeOf(BuildArgsOrString{}) {
-		return transformBuildArgsOrString()
+// Transformer returns custom merge logic for HealthCheckArgsOrString's fields.
+func (t healthCheckArgsOrStringTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ != reflect.TypeOf(HealthCheckArgsOrString{}) {
+		return nil
+	}
+
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(HealthCheckArgsOrString), src.Interface().(HealthCheckArgsOrString)
+
+		if srcStruct.HealthCheckPath != nil {
+			dstStruct.HealthCheckArgs = HTTPHealthCheckArgs{}
+		}
+
+		if !srcStruct.HealthCheckArgs.isEmpty() {
+			dstStruct.HealthCheckPath = nil
+		}
+
+		if dst.CanSet() { // For extra safety to prevent panicking.
+			dst.Set(reflect.ValueOf(dstStruct))
+		}
+		return nil
 	}
-	return transformBasic(typ)
 }
 
-// transformBasic implements customized merge logic for manifest fields that are number, string, bool, array, and duration.
-func transformBasic(typ reflect.Type) func(dst, src reflect.Value) error {
-	if typ.Kind() == reflect.Slice {
-		return transformSlice()
+type countTransformer struct{}
+
+// Transformer returns custom merge logic for Count's fields.
+func (t countTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ != reflect.TypeOf(Count{}) {
+		return nil
 	}
 
-	if typ.Kind() == reflect.Ptr {
-		for _, k := range basicKinds {
-			if typ.Elem().Kind() == k {
-				return transformPBasic()
-			}
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(Count), src.Interface().(Count)
+
+		if !srcStruct.AdvancedCount.IsEmpty() {
+			dstStruct.Value = nil
+		}
+
+		if srcStruct.Value != nil {
+			dstStruct.AdvancedCount = AdvancedCount{}
 		}
 
-		if typ.Elem().Kind() == reflect.Struct {
-			return transformPStruct()
+		if dst.CanSet() { // For extra safety to prevent panicking.
+			dst.Set(reflect.ValueOf(dstStruct))
 		}
+		return nil
+	}
+}
+
+type advancedCountTransformer struct{}
+
+// Transformer returns custom merge logic for AdvancedCount's fields.
+func (t advancedCountTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
```

Comment:
Im still a bit worried about this non-generic mutually exclusive transformers. Basically it means when we add new mutually exclusive fields we would need to add a new transformer function, or even when we add a new field to existing mutually exclusive combination, we must remember modifying the transformer function.

Let's take a step back 💭  Maybe we don't need to handle the mutually exclusive logic in `ApplyEnv` at all, because as a user they might find a field was unexpectedly unset. So if they specified `build` in the manifest and `location` in one of their env override, we could just error out saying they specified both fields. Then they would set no build/location fields in the manifest and only specify as needed in each env override. That would help us avoid loads of complex and repetitive logic for handling composite type and less error-prone 🤔